onnxruntime/core/optimizer/qdq_transformer/qdq_final_cleanup.cc

```diff
@@ -10,101 +10,145 @@
-      continue;  // node was removed as part of an earlier optimization
+  const auto match_first = node_sequence_type == NodeSequence::Q_DQ ? &QDQ::MatchQNode : &QDQ::MatchDQNode;
+  const auto match_second = node_sequence_type == NodeSequence::Q_DQ ? &QDQ::MatchDQNode : &QDQ::MatchQNode;
+
+  if (!match_first(first_node) ||
+      // not filtering on provider currently
+      // !graph_utils::IsSupportedProvider(first_node, compatible_execution_providers) ||
+      !optimizer_utils::CheckOutputEdges(graph, first_node, 1)) {
+    return false;
+  }
 
-    Node& q_node = *node_ptr;
-    ORT_RETURN_IF_ERROR(Recurse(q_node, modified, graph_level, logger));
+  Node& second_node = *graph.GetNode(first_node.OutputNodesBegin()->Index());
+  if (!match_second(second_node)
+      // not filtering on provider currently
+      // || !graph_utils::IsSupportedProvider(second_node, compatible_execution_providers)
+  ) {
+    return false;
+  }
 
-    if (!QDQ::MatchQNode(q_node) ||
-        // not filtering on provider currently
-        // !graph_utils::IsSupportedProvider(q_node, GetCompatibleExecutionProviders()) ||
-        !optimizer_utils::CheckOutputEdges(graph, q_node, 1)) {
-      continue;
-    }
+  if (node_sequence_type == NodeSequence::DQ_Q) {
+    // for DQ -> Q, check for constant, matching scale/ZP values
+    const auto get_constant_initializer = [&graph](const std::string& initializer_name) {
+      return graph.GetConstantInitializer(initializer_name, true);
+    };
 
-    Node& dq_node = *graph.GetNode(q_node.OutputNodesBegin()->Index());
-    if (!QDQ::MatchDQNode(dq_node) 
-        // not filtering on provider currently
-        // || !graph_utils::IsSupportedProvider(dq_node, GetCompatibleExecutionProviders())
-        ) {
-      continue;
+    if (!QDQ::IsQDQPairSupported(second_node, first_node, get_constant_initializer, graph.ModelPath())) {
+      return false;
     }
+  }
 
-    // we have Q -> DQ pair
+  // we have a node sequence to clean up
+  LOGS(logger, VERBOSE) << "Cleaning up back-to-back nodes: "
+                        << first_node.OpType() << " with name \"" << first_node.Name() << "\" and "
+                        << second_node.OpType() << " with name \"" << second_node.Name() << "\"";
 
-    // we support a DQ that produces a graph output if it has no output edges, or a DQ node with one output edge.
-    bool is_graph_output = graph.NodeProducesGraphOutput(dq_node);
-    auto edges_count = dq_node.GetOutputEdgesCount();
+  // we support a second_node that produces a graph output if it has no output edges, or a second_node with one output edge.
+  bool is_graph_output = graph.NodeProducesGraphOutput(second_node);
+  auto edges_count = second_node.GetOutputEdgesCount();
 
-    if ((is_graph_output && edges_count != 0) ||
-        (!is_graph_output && edges_count != 1)) {
-      continue;
-    }
+  if ((is_graph_output && edges_count != 0) ||
+      (!is_graph_output && edges_count != 1)) {
+    return false;
```

Comment:
Just curious: Why does the optimizer take a dependency on metadata like output edge count from the second node before choosing to go ahead with its finalizing logic ?

Reply:
I don't see why it couldn't also handle more than one output edge. Perhaps it was just for simplicity. Can look into supporting more.

Reply:
Mainly for simplicity. If there is more than one edge the QDQ node groups would overlap and we'd handle the model badly (so we should update the model elsewhere to address that instead of handling it here).

Reply:
Simpler also means a smaller binary size. Unless we see real models with edge cases like multiple output edges or an output edge as well as a graph output, there's potentially zero benefit from having a larger implementation that handles all these extra complexities, but a definite cost in terms of binary size, and code to write/test/maintain.